h2oai/h2o-llmstudio
# UI Automation tests for LLM Studio

### Framework
We have used Python BDD Framework with playwright.

### Tree
```
llm_studio/
|-- tests/
|   |-- ui/
|       |-- __init__.py
|       |-- conftest.py
|       |-- utils.py
|       |-- llm_studio_page.py
|       |-- llm_studio.feature
|       |-- test.py
```

### How to run the tests
- run the latest version for llm studio app on cloud platforms like https://internal-genai.dedicated.h2o.ai/
- in the console export following environment variables:
  ```bash
  export APP_URL=
  export OKTA_USER=
  export OKTA_PASSWORD=
  ```
- setup environment
  ```bash
  make setup-ui-test
  ```
- run the tests
  ```bash
  make test-ui
  ```

> Please note that this testing framework requires an instance to running prior. To integrate this test suite to CI, I will need your help.

### Important Points for Discussion:
- For now I have not added separate logging to the codebase. Should I add logging? @mnezh 
- Theres a method in playwright called `get_by_test_id` which helps testing by test ids ie:
   - for http elements like:
      ``` http
      <button data-testid="directions">Itinéraire</button>
      ```
    - we can simply do:
      ```
      page.get_by_test_id("directions").click()
      ```
    - However we dont have `data-testid=` in UI codebase. We instead have `data-test`.
    - The good news is there is a method to tackle this https://playwright.dev/python/docs/locators#set-a-custom-test-id-attribute But due to lack of documentation I wasnt able to use it
    - So I created my own method by the same name.
    - Can you please help me tackle this? @mnezh 

- Should I also add piplock to the repository? @maxjeblick
